src/sources/kafka.rs

```diff
@@ -255,16 +295,21 @@ async fn parse_message(
 // Turn the received message into a stream of parsed events.
 fn parse_stream<'a>(
     msg: &BorrowedMessage<'a>,
-    decoder: Decoder,
+    decoder: &Decoder,
     keys: Keys<'a>,
+    topics: &Topics,
 ) -> Option<(usize, impl Stream<Item = Event> + 'a)> {
     let payload = msg.payload()?; // skip messages with empty payload
 
     let rmsg = ReceivedMessage::from(msg);
 
+    if topics.failed.contains(&rmsg.topic) {
+        return None;
+    }
```

Comment:
This should be at the very top of the function, it seems like.

If we know we're going to skip any queued messages that we hadn't yet drained when marking a topic as failed, we might as well be as efficient as possible about it.

Reply:
You are right. However, there is a cost there of calling `msg.topic.to_string()` twice when there are no rejected messages. That's probably smaller, though.